### PR TITLE
Simplify logic for identifying type conflicts

### PR DESCRIPTION
**Pull request by thatplguy**
The logic for deciding when to introduce a `OneOf` relied on removing parts of `Data` objects before comparing them, which was subtle and prone to introducing conflicts if the IR changes in the future.

This PR refactors the logic and adds comments.  The semantics are the same.

### CODE REVIEW

File: spec_util/meld.go [hunk 1]
```diff
@@ -308,50 +308,37 @@ func dataEqual(dst, src *pb.Data) bool {
...
-		assignDataFormats(src, srcDataFormats)
-		assignDataFormats(dst, dstDataFormats)
-
-		// Reinstate original example values
-		src.ExampleValues = srcExampleValues
-		dst.ExampleValues = dstExampleValues
+	cmpDst := &pb.Primitive{Value: dst.Value, FormatKind: dst.FormatKind}
+	cmpSrc := &pb.Primitive{Value: src.Value, FormatKind: src.FormatKind}
+	return proto.Equal(cmpDst, cmpSrc)
+}
```
Comment (liujed) [inline spec_util/meld.go:321]: An unformatted int should subsume a formatted int, and should therefore be compatible. Similarly for other types, of course.

File: spec_util/meld.go [hunk 2]
```diff
@@ -614,3 +580,34 @@ func (m *melder) meldList(dst, src *pb.List) error {
...
+	for k := range src.Formats {
+		mergedDataFormats[k] = true
+	}
+	for k := range dst.Formats {
+		mergedDataFormats[k] = true
+	}
+	if len(mergedDataFormats) > 0 {
+		dst.Formats = mergedDataFormats
+	}
+}
```
Comment (liujed) [inline spec_util/meld.go:613]: When merging a formatted int with an unformatted int, the result should be an unformatted int. This appears to produces a formatted int. Similarly for other types, of course.

**Review by liujed — COMMENTED** [review]
The handling of unformatted primitives doesn't look quite right.

File: spec_util/meld.go (hunk 1 above)
Comment (thatplguy) [reply spec_util/meld.go:321]: Hm, true.  I don't think the existing code does that, though.  If you don't mind, I'll land this PR with just the refactoring and then do another to properly move up the type lattice during melding.

Comment (thatplguy) [reply spec_util/meld.go:321]: I'll add some test cases to check this behavior first though.

File: spec_util/meld.go (hunk 2 above)
Comment (thatplguy) [reply spec_util/meld.go:613]: As it stands, primitives are only merged if the base type and format kind are equal; the cases you're talking about currently result in a type conflict.

File: spec_util/meld.go (hunk 1 above)
Comment (thatplguy) [reply spec_util/meld.go:321]: Added tests confirming that the existing behavior does not merge up the type lattice: https://github.com/akitasoftware/superstar/pull/636

**Review by liujed — APPROVED** [review]

**Pull request merged**